qir/qat/PreTransformValidation/PreTransformValidationPass.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "PreTransformValidation/PreTransformValidationPass.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+#include <fstream>
+#include <iostream>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    PreTransformValidationPass::PreTransformValidationPass(
+        PreTransformValidationPassConfiguration const& cfg,
+        ILoggerPtr const&                              logger)
+      : config_{cfg}
+      , logger_{logger}
+    {
+    }
+
+    llvm::PreservedAnalyses PreTransformValidationPass::run(llvm::Module& module, llvm::ModuleAnalysisManager& /*mam*/)
+    {
+        uint64_t function_count{0};
+
+        if (config_.requireStraightLineCode())
+        {
+            for (auto& function : module)
+            {
+
+                if (function.isDeclaration())
+                {
+                    continue;
+                }
+
+                ++function_count;
+
+                if (function_count > 1)
+                {
+                    if (logger_)
+                    {
+                        logger_->setLocationFromValue(&function.getEntryBlock());
+                        logger_->error("Expected straight line code, but multiple functions present.");
+                    }
```

Comment:
I have an impression that line 45 can be executed multiple times in the loop (starting in line 30). Did you mean placing `break;` or `return;` between lines 45 and 46?

Same for line 62 below.

Reply:
The reason for this is that if a logger is present, we highlight all errors. In this case, that means all functions beyond the first one. However, in the absence of a logger, we throw to 1) create a signal that the program has failed and 2) to avoid flooding the `stderr`